File: array_string/sliding_window/_go/maximum_sub_array.go
package main

import "fmt"

func maxSubArray(nums []int) int {
	mxSum := nums[0]
	start := 0
	currentSum := nums[0]
	result := []int{nums[0]}
	for end := 1; end < len(nums); end++ {
		if nums[end] > currentSum+nums[end] {
			currentSum = nums[end]
			start = end
		} else {
			currentSum += nums[end]
		}
		if currentSum > mxSum {
			mxSum = currentSum
			result = nums[start : end+1]
		}
	}
	return Sum(result)
}

func Sum(result []int) int {
	val := 0
	for _, v := range result {
		val += v
	}
	return val
}
func main() {
	nums := []int{-2, 1, -3, 4, -1, 2, 1, -5, 4}
	fmt.Println(maxSubArray(nums))
}

File: array_string/sliding_window/maximum_sub_array.py
# https://leetcode.com/problems/maximum-subarray/

def max_subarray_sum(nums):
    max_sum = nums[0]
    result = [nums[0]]
    start = 0
    end = 1
    current_sum = nums[0]
    for end in range(1, len(nums)):
        # If current value greater than subarray sum
        # That means there is another greatest subarray
        if nums[end] > current_sum + nums[end]:
            current_sum = nums[end]
            start = end
        else:
            # If current value not greater than subarray sum
            # That means we need to extend current subarray
            current_sum += nums[end]

        if current_sum > max_sum:
            max_sum = current_sum
            result = nums[start:end+1]
    return result

# nums = [5, 4, -1, 7, 8]
# nums = [1,-1]
# nums =  [-2,1,-3,-55, 4444]
nums = [-2,-1]
print("Largest subarray sum:", max_subarray_sum(nums))
